host.py

```diff
@@ -1,24 +1,85 @@
 import json
+import logging
 import os
 import shlex
 import subprocess
 import sys
+import threading
 import typing
 
 from abc import ABC
 from abc import abstractmethod
+from collections.abc import Iterable
+from collections.abc import Mapping
 from dataclasses import dataclass
 from typing import Any
+from typing import Optional
 
 from logger import logger
 
 
+_lock = threading.Lock()
+
+_global_counter_val = 0
+
+
+def _global_counter() -> int:
+    # For each run() call, we log a message when staring the command and when
+    # completing it. Add a unique number to those logging statements, so that
+    # we can easier find them in a large log.
+    with _lock:
+        global _global_counter_val
+        _global_counter_val += 1
+        return _global_counter_val
+
+
+T = typing.TypeVar("T", bound=str | bytes)
+
+
 @dataclass(frozen=True)
-class Result:
-    out: str
-    err: str
+class BaseResult(ABC, typing.Generic[T]):
+    out: T
+    err: T
     returncode: int
 
+    @property
+    def success(self) -> bool:
+        return self.returncode == 0
+
+    def debug_str(self) -> str:
+        if self.returncode == 0:
+            status = "succcess"
+        else:
+            status = f"failed ({self.returncode})"
+
+        out = ""
+        if self.out:
+            out = f"; out={repr(self.out)}"
+
+        err = ""
+        if self.err:
+            err = f"; err={repr(self.err)}"
+
+        return f"{status}{out}{err}"
+
+    def debug_msg(self) -> str:
+        return f"cmd {self.debug_str()}"
+
+
+@dataclass(frozen=True)
+class Result(BaseResult[str]):
+    pass
+
+
+@dataclass(frozen=True)
+class BinResult(BaseResult[bytes]):
```

Comment:
What is an example of a binary result?

Reply:
there is one in the commit message:
```
python -c 'import host; r = host.local.run("printf \"foo:\\705\""); print(r)'
```
or really anytime, you get binary result, like 
```
$ python -c 'import host; host.local.run("echo -n hello | gzip")'
2024-06-18 19:47:05 DEBUG   [th:140307334326080]: cmd[1;localhost]: call `echo -n hello | gzip`
2024-06-18 19:47:05 ERROR   [th:140307334326080]: cmd[1;localhost]: └──> `echo -n hello | gzip`: [BINARY] [UNEXPECTED_BINARY] succcess; out=b'\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03\xcbH\xcd\xc9\xc9\x07\x00\x86\xa6\x106\x05\x00\x00\x00'
```

Since we call other processes (that we don't trust), we should not assume that we get benign UTF-8.

also, CDA has `Host.read_file()` which is implemented as `f"cat {filename}"`. That cannot read binary files.

Usually, we don't call programs that output binary data. The main point here is the logging, to catch cases where we get unexpected binary output.

Reply:
Ah i see. It is to catch unexpected cases.